internal/metastore/kv/datacoord/kv_catalog.go

```diff
@@ -102,23 +104,37 @@ func (kc *Catalog) AlterSegments(ctx context.Context, newSegments []*datapb.Segm
 	if len(newSegments) == 0 {
 		return nil
 	}
-
-	kvs := make(map[string]string)
+	kvsBySeg := make(map[int64]map[string]string)
 	for _, segment := range newSegments {
 		segmentKvs, err := buildSegmentAndBinlogsKvs(segment)
 		if err != nil {
 			return err
 		}
-		maps.Copy(kvs, segmentKvs)
-	}
-
-	saveFn := func(partialKvs map[string]string) error {
-		return kc.Txn.MultiSave(partialKvs)
+		kvsBySeg[segment.GetID()] = make(map[string]string)
+		maps.Copy(kvsBySeg[segment.GetID()], segmentKvs)
+	}
+	// Split kvs into multiple operations to avoid over-sized operations.
+	// Also make sure kvs of the same segment are not split into different operations.
+	kvsPiece := make(map[string]string)
+	currSize := 0
+	for _, kvs := range kvsBySeg {
+		if currSize+len(kvs) >= maxEtcdTxnNum {
+			if err := kc.Txn.MultiSave(kvsPiece); err != nil {
+				log.Error("failed to call MultiSave", zap.Error(err))
+				return err
+			}
+			kvsPiece = make(map[string]string)
+			currSize = 0
+		}
+		maps.Copy(kvsPiece, kvs)
+		currSize += len(kvs)
 	}
```

Comment:
This piece of logic is implemented inside `etcd.SaveByBatch`

Reply:
Yes but not by segment, so there's a chance that kvs of a same segment are split into different operations.

Reply:
If the transaction number of one segment is greater than 64, it still needs to be split into multiple operation.

Reply:
In this case we won't have more than 64 kvs (ops) for a single segment.

Reply:
If collection has more than  32 fields, it shall have 32*2+1 ops for save one segment

Reply:
Ah I see what you mean, that's true. If that case these binlog kvs could still get separated into multiple OPs.